core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java

```diff
@@ -154,13 +154,13 @@ public Schema.Field field(Schema.Field field, Supplier<Schema> fieldResult) {
 
   @Override
   public Schema union(Schema union, Iterable<Schema> options) {
-    Preconditions.checkState(AvroSchemaUtil.isOptionSchema(union),
-        "Invalid schema: non-option unions are not supported: %s", union);
-    Schema nonNullOriginal = AvroSchemaUtil.fromOption(union);
-    Schema nonNullResult = AvroSchemaUtil.fromOptions(Lists.newArrayList(options));
+    if (AvroSchemaUtil.isOptionSchema(union)) {
+      Schema nonNullOriginal = AvroSchemaUtil.fromOption(union);
+      Schema nonNullResult = AvroSchemaUtil.fromOptions(Lists.newArrayList(options));
 
-    if (!Objects.equals(nonNullOriginal, nonNullResult)) {
-      return AvroSchemaUtil.toOption(nonNullResult);
+      if (!Objects.equals(nonNullOriginal, nonNullResult)) {
+        return AvroSchemaUtil.toOption(nonNullResult);
+      }
     }
 
     return union;
```

Comment:
Thanks for catching this. Improved implementation in complex union case. Please take a look.

Reply:
Can we add a test case for these scenarios? Given that the issue here is nuanced and not already being tested elsewhere, we should add test cases to avoid future regressions.

Reply:
Added unit tests as suggested in `TestBuildAvroProjection`.